### PR TITLE
Hs 3204

### PR DESCRIPTION
**Pull request by palegre-fluendo**
The branch allows behave to run each feature in a different process, basically it uses fork() to create a child process to run the feature and receives the result using a pipe.
It also controls a tag timeout to kill the child process if it tokes too long, so now we are protected against deadlocks in the tests.

### CODE REVIEW

File: behave/runner.py [hunk 1]
```diff
@@ -21,13 +21,16 @@
...
 
 if six.PY2:
     # -- USE PYTHON3 BACKPORT: With unicode traceback support.
     import traceback2 as traceback
 else:
     import traceback
-
+    
+import time
+import signal
```
Comment (ylatuya) [inline behave/runner.py]: The common practice is to import python modules first, see the block that goes from line 8 to 12

File: behave/runner.py [hunk 2]
```diff
@@ -495,6 +498,54 @@ def path_getrootdir(path):
     return os.path.sep
 
 
+def forEachStatusInFeatures(feature, f, param):
```
Comment (ylatuya) [inline behave/runner.py]: These methods should be private in the class, you don't want to export them in the module.

File: behave/runner.py [hunk 3]
```diff
@@ -495,6 +498,54 @@ def path_getrootdir(path):
...
+            if hasattr(s, 'scenarios'):
+                foreachscenario(s.scenarios, f, param)
+            else:
+                f(s, param)
+                for step in s.steps:
+                    f(step, param)
+
+    f(feature, param)
+    foreachscenario(feature.scenarios, f, param)
+
```
Comment (ylatuya) [inline behave/runner.py:513]: 2 line jumps for pep8 :)

File: behave/runner.py [hunk 4]
```diff
@@ -598,6 +649,69 @@ def stop_capture(self):
...
+            def read_timeout(tags):
+                tag_name = "timeout="
+                for tag in tags:
+                    res = tag.find(tag_name)
+                    if res == 0:
+                        return int(tag[len(tag_name):])
+                return 0
+
+            timeout = read_timeout(feature.tags)
+            r, w = os.pipe()
```
Comment (ylatuya) [inline behave/runner.py:671]: I am worried this implementation might not be cross-platform. Python has the multiprocess module that handles most of what is being done here.

File: behave/runner.py [hunk 5]
```diff
@@ -618,21 +732,9 @@ def run_model(self, features=None):
...
-                    if failed:
-                        failed_count += 1
-                        if self.config.stop or self.aborted:
-                            # -- FAIL-EARLY: After first failure.
-                            run_feature = False
-                except KeyboardInterrupt:
-                    self.aborted = True
-                    failed_count += 1
-                    run_feature = False
+                print("Running {}", feature)
```
Comment (ylatuya) [inline behave/runner.py]: If we want this to end up upstream we probably need this to be optional

File: behave/runner.py [hunk 6]
```diff
@@ -598,6 +649,69 @@ def stop_capture(self):
...
         self.capture_controller.teardown_capture()
 
+    def run_one_feature(self, feature):
+        try:
+            self.feature = feature
+            for formatter in self.formatters:
+                formatter.uri(feature.filename)
+
+            def read_timeout(tags):
+                tag_name = "timeout="
```
Comment (ylatuya) [inline behave/runner.py:663]: This is probably a bit Ad-hoc for our case, but right know I don't have any suggestion.
Maybe opening a PR upstream would help.

File: behave/runner.py [hunk 7]
```diff
@@ -598,6 +649,69 @@ def stop_capture(self):
...
+                for tag in tags:
+                    res = tag.find(tag_name)
+                    if res == 0:
+                        return int(tag[len(tag_name):])
+                return 0
+
+            timeout = read_timeout(feature.tags)
+            r, w = os.pipe()
+            newpid = os.fork()
```
Comment (ylatuya) [inline behave/runner.py:672]: Indeed, this isn't portable. Make sure the implementation complies with the limitations of each platform https://docs.python.org/2/library/multiprocessing.html#windows

**Review by ylatuya — CHANGES_REQUESTED** [review]

File: behave/runner.py (hunk 4 above)
Comment (palegre-fluendo) [reply behave/runner.py:671]: Ok, let me check if I can do the same with python multiprocess module.

File: behave/runner.py (hunk 6 above)
Comment (palegre-fluendo) [reply behave/runner.py:663]: Of course it is Ad-hoc for our case, in fact I was thinking more a file patch than in an upstream PR. Definitively reading tag here does not follow the original project philosophy.

File: behave/runner.py (hunk 7 above)
Comment (palegre-fluendo) [reply behave/runner.py:672]: Reviewing, thanks.

**Comment by palegre-fluendo** [comment]
Implemented using python multiprocessing in [HS-3024-2](https://github.com/fluendo/behave/pull/2).

**Pull request closed without merging**